Update Assets Groups

jorgemd24 opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes part of https://github.com/woocommerce/google-listings-and-ads/issues/1780

This PR allows updating the Asset Groups and their assets. The fields that are allowed to be updated in the asset group are:

- The Final URL - The Asset Group final URL.
- Path 1 - The first part of the text that may appear appended to the URL displayed in the ad.
- Path 2 - Second part of the text that may appear appended to the URL displayed in the ad. This field can only be set when path1 is set.
- And their assets.

Assets are [immutable](https://github.com/googleads/google-ads-php/blob/main/src/Google/Ads/GoogleAds/V11/Resources/Asset.php#L22-L25) and are linked to the Asset Group through the `AssetGroupAsset`, this means that in the case of updating an asset it is required to create a new `Asset` and its link in the `AssetGroupAsset`. 

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

### Get an Asset Group ID

1. As the branch `feature/pmax-assets` is using some new imports from the GoogleAds Library, you may need to remove your vendor folder `rm -Rf vendor/*` and install again the dependencies using `composer install`.
2. You can get an asset group id for a specific campaign by making a request to  `/gla/ads/campaigns/YOUR_CAMPAIGN_ID/asset-groups` and get the `id` from the response.


### Update Asset Group

1.  Make a request: `PUT gla/ads/campaigns/asset-groups/Your_AssetGroupID` with a JSON body.  The JSON body can be formed with the following fields:
```
{
"final_url": string, // The final URL and shopping merchant URL needs to have the same domain.
"path1": string,
"path2": string,
"assets": [
   {
      "id": string|null,
      "field_type":  string. // See enums values: https://github.com/woocommerce/google-listings-and-ads/pull/1832/files#diff-387aaa2a15a475b5e67f27d02bb3c638ca8c574f914a880bb4e95b54ce0a751cR236-R245
      "content": string|null. // If the field type is logo|square_marketing_image|marketing_images should be the url.
   }, 
   .... More assets
]
```
- If the `id` is null, it is considered as a new asset.
- If the `content` is null, it is considered as a deleted asset.
- If `id` and `content` are present, the asset will be updated.

4. Check that the Asset Group has been updated. You can check it in the Google Ads interface or call the endpoint: `/gla/ads/campaigns/YOUR_CAMPAIGN_ID/asset-groups`.


### Additional details:
- If there is an image with similar content loaded in the asset group previously, the Ads API will skip it.
- I realize that WCS has a maximum of 1 MB for POST requests, so when we are uploading multiple images, it will fail. The solution is increasing the maximum payload by adding the param `maxBytes` in the following line: https://github.com/Automattic/woocommerce-connect-server/blob/3f51d1ba3b90145137de5390366619621cfcd989/lib/google-proxy/index.js#L56-L58 . I will create a PR to increase this payload size.
- If you are updating images such as `logo`, `square_marketing_image` or `marketing_image` you need to use the right size image. You can use the endpoint `gla/assets/suggestions?id=POST_ID|TERM_ID&type=post|term` and it will suggest images with the right size if the post or the category has images.
- The image URL needs to be accessible. Localhost links may not work.
- See Assets requirements here: https://developers.google.com/google-ads/api/docs/performance-max/assets
<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

jorgemd24 committed:
Add edit asset group endpoint

jorgemd24 committed:
Add create operation asset

jorgemd24 committed:
Add create operation AdsAssetGroupAsset

jorgemd24 committed:
Add delete operation AssetGroupAsset

jorgemd24 committed:
Add mutate operation asset group asset

jorgemd24 committed:
Edit Asset group assets

jorgemd24 committed:
Add temporary id asset group asset

jorgemd24 committed:
Add edit_asset_group

jorgemd24 committed:
Refactor edit_assets_group_assets

jorgemd24 committed:
Add get_asset_type_by_field_type

jorgemd24 committed:
Add Catch ApiException

jorgemd24 committed:
Add edit asset group

jorgemd24 committed:
Add asset group response

jorgemd24 committed:
Rename route asset-group

jorgemd24 committed:
Add tests AssetGroup

jorgemd24 committed:
Update asset group controller

jorgemd24 committed:
Add call to action type

jorgemd24 committed:
Rename function create_operation

jorgemd24 committed:
Add tests for Asset Group Assets operations

jorgemd24 committed:
Add Assets tests

jorgemd24 committed:
Handle wp error if the image url is not valid

jorgemd24 committed:
Update is_wp_error image data

jorgemd24 committed:
Add tests for editing asset group controller

jorgemd24 committed:
Update PHPDocs

jorgemd24 committed:
Add enums AssetGroup controller

jorgemd24 committed:
Add final_url property

jorgemd24 committed:
Refactor AdsAsset

jorgemd24 committed:
Get assets by URL

jorgemd24 committed:
Add CONTAINS ANY to the query

jorgemd24 committed:
Update AssetSuggestionsService

jorgemd24 committed:
Merge branch 'feature/pmax-assets' into add/fetch-assets-using-final-url

jorgemd24 committed:
Combine results wp_assets and asset group assets

jorgemd24 committed:
Update phpdocs

jorgemd24 committed:
Simplify edit asset group params

jorgemd24 committed:
Rename function get_assets_by_final_url

jorgemd24 committed:
Refactor combine results function

jorgemd24 committed:
Add tests combine asset results

jorgemd24 committed:
Search url with and without trailing slash

jorgemd24 committed:
Refactor get_assets_by_final_url

puntope commented on src/API/Google/AdsAssetGroupAsset.php:
```diff
@@ -121,4 +133,84 @@ public function get_assets_by_asset_group_ids( array $asset_groups_ids ): array
+	 *
+	 * @return array The asset group asset operations.
+	 */
+	public function edit_operations_assets_group_assets( int $asset_group_id, array $assets ): array {
```
💅 

What do you think about renaming this as `edit_operations`? Otherwise we can rename  `create_operation` as `create_operation_assets_group_assets` and `delete_operation` as `delete_operation_assets_group_assets`.

In brief I was expecting some consistency in naming.

jorgemd24 replied:
That sounds right, I have changed it to `edit_operations`. See: [374fdf8](https://github.com/woocommerce/google-listings-and-ads/pull/1832/commits/374fdf8888da4a47c232af4e0a6a3f15273265e9)

puntope commented on src/API/Google/AdsAssetGroupAsset.php:
```diff
@@ -121,4 +133,84 @@ public function get_assets_by_asset_group_ids( array $asset_groups_ids ): array
+	}
+
+	/**
+	 * Returns a campaign delete operation.
```
❓  Shouldn't this be "Returns a delete operation for asset group asset"?

puntope commented:
❓ 

When I tried to create a new Call to Action I got 500 error

```json
{
"final_url": "https://805e-136-144-33-28.ngrok.io",
"path1": "test1",
"path2": "test2",
"assets": [
   {
      "id": null,
      "field_type": "call_to_action_selection",
      "content": "test"
   }
]
}
```

jorgemd24 committed:
Fix issue with status mapping::number, if the label is not found is return `false` but the function signature it says that should return an int value

jorgemd24 committed:
Handle clase when CallToActionType is undefined

jorgemd24 committed:
Adjust PHPDoc

jorgemd24 committed:
Rename edit_operations_asset_group_operations

jorgemd24 commented:
Thanks, @puntope for having a look at this.

> When I tried to create a new Call to Action I got 500 error.

The issue was caused because `array_search` returns a boolean if the item is not found, in this case `test` is not a valid value for the call_to_action field. I fix the issue here: [9a24355](https://github.com/woocommerce/google-listings-and-ads/pull/1832/commits/9a243553fa7c8862cc5a7ca5c7f8c1459184843f)

You can find the possible values for `call_to_action` here: https://github.com/woocommerce/google-listings-and-ads/blob/374fdf8888da4a47c232af4e0a6a3f15273265e9/src/API/Google/CallToActionType.php#L34-L95.

jorgemd24 committed:
Merge branch 'add/update-assets-groups' into add/fetch-assets-using-final-url

jorgemd24 committed:
Merge branch 'feature/pmax-assets' into add/update-assets-groups

jorgemd24 committed:
Merge branch 'add/update-assets-groups' into add/fetch-assets-using-final-url

jorgemd24 committed:
Only suggest WP assets if there are not assets available from the asset group

jorgemd24 committed:
Add default CALL_TO_ACTION_SELECTION

jorgemd24 committed:
Add path1 and path2 to the asset suggestion final url

jorgemd24 committed:
Handle case if the final url is not found

jorgemd24 committed:
Adjust test get final url

jorgemd24 committed:
update field_type schema as required

puntope commented:
Hello @jorgemd24 I was testing again the call to action type but still I'm not able to do it. See screenshot. What I'm doing wrong? 

<img width="1398" alt="Screenshot 2023-01-04 at 17 27 11" src="https://user-images.githubusercontent.com/5908855/210542387-6ba1cb00-ca7c-4f64-a0dc-0b5cbc970d26.png">

jorgemd24 commented:
@puntope,

The message is referring that the maximum number of assets for that field type (`call_to_action_selection`) is exceeded.  In this case, the maximum number for `call_to_action_selection` assets  is `1`, so probably the asset group that you are trying to update already has one `call_to_action_selection`, therefore is failing. Once is created you can update/delete it using the asset id. If you delete it then you will be able to create it again.

jorgemd24 committed:
Refactor asset suggestion to return the assets from the asset group

jorgemd24 committed:
Update tests

jorgemd24 committed:
Add missing type

jorgemd24 committed:
Add indents sprintf

jorgemd24 committed:
Use trailingslashit

jorgemd24 committed:
Add only_first_asset_group

puntope commented on src/API/Google/AdsAsset.php:
```diff
@@ -18,7 +26,97 @@
+	 *
+	 * @return string The Asset resource name.
+	 */
+	protected function temporary_resource_name( $temporary_id = self::TEMPORARY_ID ): string {
```
💅  We can add the int type here

puntope commented on src/API/Site/Controllers/Ads/AssetGroupController.php:
```diff
@@ -42,6 +43,17 @@ public function __construct( RESTServer $rest_server, AdsAssetGroup $ads_asset_g
 	 */
 	public function register_routes(): void {
+		$this->register_route(
+			'ads/campaigns/asset-groups/(?P<id>[\d]+)',
```
❓   Why here in the PUT the ID param is at the end and in the GET is before the asset-groups URL segment?

jorgemd24 replied:
In the `PUT` the `id` is the `asset group id` and in the `GET` the `id` is the `campaign id`, however, I am considering renaming the `GET` endpoint ( in a different PR), so instead of using  `ads/campaigns/CAMPAIGN_ID/asset-groups` I think it is more convenient something like: `ads/campaigns/asset-groups?campaign_id=123445`, WDYT?

puntope replied:
Ahhh gotcha. I missed that. For me both options are valid tho.

puntope commented on src/API/Site/Controllers/Ads/AssetGroupController.php:
```diff
@@ -42,6 +43,17 @@ public function __construct( RESTServer $rest_server, AdsAssetGroup $ads_asset_g
+					'methods'             => TransportMethods::EDITABLE,
+					'callback'            => $this->edit_asset_group_callback(),
+					'permission_callback' => $this->get_permission_callback(),
+					'args'                => $this->get_edit_params(),
```
💅  For consistency maybe we could rename this to `edit_asset_group_params`

puntope commented on src/API/Site/Controllers/Ads/AssetGroupController.php:
```diff
@@ -56,6 +68,50 @@ public function register_routes(): void {
+	/**
+	 * Get the edit params to update an asset group.
+	 */
+	public function get_edit_params() {
```
💅  Missed return type

puntope commented on src/API/Google/AdsAssetGroupAsset.php:
```diff
@@ -121,4 +133,84 @@ public function get_assets_by_asset_group_ids( array $asset_groups_ids ): array
+	 * @param int   $asset_group_id The asset group id.
+	 * @param array $assets The assets to create.
+	 *
+	 * @return array The asset group asset operations.
```
💅  Seems like `create_operation` could potentially throw an extension which is not handled neither added to the phpdocs.

jorgemd24 replied:
I am not sure about this, what I can see the `MutateOperation` or any of their methods `setAssetGroupAssetOperation `or `setCreate`, would not throw an exception. The exception will throw when the mutating operations are executed in this line:  https://github.com/woocommerce/google-listings-and-ads/blob/1d9f8deda7f031738c8fa457948a65fae99a7ece/src/API/Google/AdsAssetGroup.php#L311

puntope replied:
What I meant is that inside `$this->asset->create_operation(`) we throw an exception in the default method. 

```
default:
			throw new Exception( 'Asset type not supported' );
```

This exception is not being handled inside `edit_operation` method. So either we handled that exception inside `edit_operation` or we can just add  `* @throws Exception` in PHPDOC

puntope commented on src/API/Google/AdsAssetGroupAsset.php:
```diff
@@ -44,6 +48,14 @@ class AdsAssetGroupAsset implements OptionsAwareInterface {
 	 */
```
In this file

`use Automattic\WooCommerce\GoogleListingsAndAds\API\Google\AdsAsset;`
Is not being used.

puntope commented on src/API/Google/AdsAssetGroupAsset.php:
```diff
@@ -121,4 +133,84 @@ public function get_assets_by_asset_group_ids( array $asset_groups_ids ): array
+		}
+
+		// Delete asset group assets operations must be executed last so we are never under the minimum quantity.
+		$operations = array_merge( $assets_operations, $asset_group_assets_operations, $delete_asset_group_assets_operations );
+
+		return $operations;
```
💅  We can just return `array_merge( $assets_operations, $asset_group_assets_operations, $delete_asset_group_assets_operations )`

puntope commented on src/API/Google/AdsAssetGroupAsset.php:
```diff
@@ -121,4 +133,84 @@ public function get_assets_by_asset_group_ids( array $asset_groups_ids ): array
+	 *
+	 * @return AssetGroupAssetOperation The create operation for the asset group asset.
+	 */
+	protected function create_operation( int $asset_group_id, string $asset_field_type, int $asset_id ): MutateOperation {
```
❓  In the type it says MutateOperation, in the PHPDOC AssetGroupAssetOperation

puntope commented on src/API/Google/AdsAsset.php:
```diff
@@ -18,7 +26,97 @@
+				$asset->setCallToActionAsset( new CallToActionAsset( [ 'call_to_action' => CallToActionType::number( $data['content'] ) ] ) );
+				break;
+			case AssetType::IMAGE:
+				$image_data = wp_remote_get( $data['content'] );
+
+				if ( is_wp_error( $image_data ) || empty( $image_data['body'] ) ) {
+					throw new Exception( 'Incorrect image asset url.' );
+				}
```
❓  Wondering if we can wrap this in the WP Proxy

jorgemd24 replied:
I have added `wp_remote_get` to the WP proxy as it makes easier the testing for the Image assets, however, I am not sure if we should include throwing the `exception`  in the WP Proxy as it will depend on the consumer how it wants to treat the case where exists a `wp_error` or if we want to consider an empty body as an `Exception`. See [b7664f0](https://github.com/woocommerce/google-listings-and-ads/pull/1832/commits/b7664f060f631c6f110959f10667e94b7ad314f1)

puntope commented on src/API/Google/AdsAssetGroup.php:
```diff
@@ -283,6 +284,69 @@ protected function get_assets( array $asset_groups ): array {
 		return $asset_groups;
 	}
```
💅  `use Automattic\WooCommerce\GoogleListingsAndAds\API\Google\AdsAssetGroupAsset;`

Not being used

puntope commented on src/API/Google/AdsAssetGroup.php:
```diff
@@ -283,6 +284,69 @@ protected function get_assets( array $asset_groups ): array {
+				$operations[] = $this->edit_operation( $asset_group_id, $data );
+			}
+
+			if ( ! empty( $operations ) ) {
+				$this->client->getGoogleAdsServiceClient()->mutate( $this->options->get_ads_id(), $operations );
+			}
+
+			return $asset_group_id;
```
❓ I take as granted that if the client request is not successful it will trow exception?

jorgemd24 replied:
Correct 👍

puntope commented on tests/Unit/API/Google/AdsAssetTest.php:
```diff
@@ -0,0 +1,141 @@
+			'content'    => 'Test headline',
+			'id'         => self::TEST_ASSET_ID,
+		];
+		$row  = $this->generate_asset_row_mock( $data );
+
+		$expected = [
+			'id'      => self::TEST_ASSET_ID,
+			'content' => $data['content'],
+		];
+
+		$this->assertEquals( $expected, $this->asset->convert_asset( $row ) );
```
💅  Minor since it's a test. But potentially we could refactor these as 

`assertAssetTypeConversion( $type, $data)`

puntope commented on tests/Unit/API/Google/AdsAssetTest.php:
```diff
@@ -0,0 +1,141 @@
+
+	}
+
+	public function test_create_operation_call_to_action_asset() {
```
❓ I guess we missed here Image type asset create operation test?

puntope reviewed (commented):
Howdy @jorgemd24 

Behaviour looking good. I did finally some deeper checks in regards code. Mostly minor. 

Thanks in advance.

jorgemd24 committed:
Add missing type temporary_resource_name

jorgemd24 committed:
Remove unnesary imports and adjust phpdocs

jorgemd24 committed:
Rename edit_asset_group_params

jorgemd24 committed:
Refactor tests

jorgemd24 committed:
Fix wrong phpdoc space

jorgemd24 committed:
Add wp_remote to WP proxy

jorgemd24 committed:
Add test image create operation

jorgemd24 commented:
Thanks, @puntope for having a look at it. 

I have addressed your comments and replied to your questions. I guess this is ready for another round! 

Thanks!

puntope reviewed (approved):
✅  Thanks for the changes @jorgemd24 

This is looking good for me

jorgemd24 committed:
Add throws exception PHPDoc

jorgemd24 committed:
Merge branch 'add/update-assets-groups' into add/fetch-assets-using-final-url

jorgemd24 committed:
Merge pull request #1837 from woocommerce/add/fetch-assets-using-final-url

Fetch assets from other Asset Groups using the final url
